lib/naxum/src/serve.rs

```diff
@@ -176,32 +177,30 @@ where
                         tracker.close();
                         break;
                     }
-                    (msg, permit) = next_message(&mut stream, semaphore.clone(), failed_count) => {
+                    (msg, permit) = next_message(&mut stream, semaphore.clone()) => {
                         match msg {
                             Some(Ok(msg)) => {
-                                failed_count = 0;
-                                metric!(counter.naxum.failed_count = 0);
-
+                                metric!(counter.naxum.next_message.failed = 0);
+                                metric!(counter.naxum.next_message.none = 0);
                                 (msg, permit)
                             },
                             Some(Err(err)) => {
-                                // TODO(fnichol): this level might need to be `trace!()`, just
-                                // unclear at the moment
-                                warn!(si.error.message = ?err, "failed to read next message from stream");
-                                failed_count += 1;
+                                error!(si.error.message = ?err, "failed to read next message from stream");
+                                metric!(counter.naxum.next_message.failed = 1);
                                 continue;
                             },
                             None => {
-                                info!("stream is closed, breaking out of loop");
-                                tracker.close();
-                                break;
+                                warn!(sleep_duration_ms = ?NEXT_NONE_MESSAGE_SLEEP_DURATION, "next message is 'None' on provided stream: going to sleep");
+                                metric!(counter.naxum.next_message.none = 1);
+                                std::thread::sleep(NEXT_NONE_MESSAGE_SLEEP_DURATION);
```

Comment:
Non-async sleep... I'm a bit iffy on this. We have never seen this code exercised. This follows the recommendation to accept `Ready(None)` when the connection is unhealthy and reconnection is happening.

Reply:
We can't do this--the Tokio worker thread will be blocked here for this duration and nothing on this worker will make progress, including any shutdown logic.

Reply:
Removed!